### ISSUE TITLE
route decorator on MethodView classes

### ISSUE THREAD

**Issue by dougthor42** [pallets/flask#3404]
### TL;DR:
+ Being able to use a Blueprint's `@pb.route` decorator on a `MethodView` class can help made code more readable
+ It should be a pretty trivial fix.
+ I'm working on tests and a merge request right now.

### Expected Behavior
Users should be able to combine blueprints with `MethodView`.

```python
from flask import Blueprint
from flask.views import MethodView

api = Blueprint("api", __name__, url_prefix="/api/v1")

@api.route("/users")
@api.route("/users/<user_id>")
class UserAPI(MethodView):

    def get(self, user_id=None):
        if user_id is None:
            return "list of all users"
        else:
            return f"just user {user_id}"

    def post(self, user_id):
        return f"POST {user_id}"

    def delete(self, user_id):
        return f"DELETE {user_id}"

    def put(self, user_id):
        return f"PUT {user_id}"

    def PATCH(self, user_id):
        return f"PATCH {user_id}"
```

### Actual Behavior
Using a blueprint's `@route` decorator on a `MethodView` results in `TypeError` is raised:

```pytb
Traceback (most recent call last):
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    return self.finalize_request(rv)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1967, in finalize_request
    response = self.make_response(rv)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2130, in make_response
    " {rv.__class__.__name__}.".format(rv=rv)
TypeError: The view function did not return a valid response. The return type must be a string, dict, tuple, Response instance, or WSGI callable, but it was a UserAPI.
```

Instead, users need to manually add rules to the app after the fact, which can be hard to read and requires that the `app` instance be available in the module.

```python
user_view = UserAPI.as_view('user_api')
app.add_url_rule('/users/', defaults={'user_id': None},
                 view_func=user_view, methods=['GET',])
app.add_url_rule('/users/', view_func=user_view, methods=['POST',])
app.add_url_rule('/users/<int:user_id>', view_func=user_view,
                 methods=['GET', 'PUT', 'PATCH', 'DELETE'])
```

### Environment

* Python version: 3.6.8
* Flask version: 1.1.1
* Werkzeug version: 0.16.0

### The Fix

Based on my testing is *should* be relatively trivial to support this. All that needs to be done is edit `Blueprint.route`:

```diff
diff --git a/src/flask/blueprints.py b/src/flask/blueprints.py
index 8978104d..f6f62dd6 100644
--- a/src/flask/blueprints.py
+++ b/src/flask/blueprints.py
@@ -11,6 +11,8 @@
 """
 from functools import update_wrapper

+from .views import MethodViewType
+
 from .helpers import _endpoint_from_view_func
 from .helpers import _PackageBoundObject

@@ -276,6 +278,11 @@ class Blueprint(_PackageBoundObject):

         def decorator(f):
             endpoint = options.pop("endpoint", f.__name__)
+
+            # Support decorating MethodView classes
+            if isinstance(f, MethodViewType):
+                f = f.as_view(endpoint)
+
             self.add_url_rule(rule, endpoint, f, **options)
             return f
```

**Note:** I originally saw this feature in the `marshmallow-code/flask-smorest` library (originally `Nobatek/flask-rest-api`) [flask_smorest/blueprint.py](https://github.com/marshmallow-code/flask-smorest/blob/cc618f74abc2035d9f2904efed32cfc3bafad3ab/flask_smorest/blueprint.py#L90).
The above code is taken almost verbatim from that library.

**Comment by davidism** [comment]
Thanks for looking at this. However, I don't think I want to add this at this time.

You can already use `MethodView` with `Blueprint` with `bp.add_url_rule()`, which corresponds to `app.add_url_rule()`. I don't want to add a special case for certain types, even if they are built-in. This also obscures the behavior of `as_view()`, which is important for `View` classes. It's also not clear why this wouldn't apply to all `View` subclasses or `app.route`.

One weird trick though, if you really want to use `bp.route`:

```python
class MyView(MethodView):
    decorators = [..., bp.route("/my-view")]

MyView.as_view("my_view")
```

**Comment by davidism** [comment]
Eh, on second thought using a view class without any arguments is probably pretty common, so maybe this would be good, although I'd want to make it more general and make the docs clear on when to use each form. @dougthor42 @ThiefMaster thoughts?

**Comment by davidism** [comment]
One downside is that if the class really does require arguments, you'll get a somewhat confusing error if you try to use `route`, and the message won't be able to say "you need to use `as_view`" for this view class."

**Comment by davidism** [comment]
Going to close this due to lack of response. I'm open to merging this in the future, but it needs improvement.

**Comment by dougthor42** [comment]
I guess I'm not familiar with using view classes with args. Can you provide an example of that, or tell me where in the docs to look? Are you talking about args to `__init__`? Like you have in [Pluggable Views](https://flask.palletsprojects.com/en/1.1.x/views/)?:

```
class RenderTemplateView(View):
    def __init__(self, template_name):
        self.template_name = template_name
```

I'll update the docs as best I can. Not sure what I need to do to make it more general though - I could use some guidance there.

**Comment by dfeinzeig** [comment]
@davidism , can you please point at an example of using MethodView with blueprints? i'm not understanding how the pieces fit together in that case.

**Comment by dougthor42** [comment]
@dfeinzeig You can take a look at my project [`trendlines`](https://github.com/dougthor42/trendlines), specifically [src/trendlines/routes.py `DataByName`](https://github.com/dougthor42/trendlines/blob/master/src/trendlines/routes.py#L133) to see how it looks after the proposed change. Note that that project is using the `marshmallow-code/flask-smorest` library which basically implements what I've proposed in this issue.

As for the current functionality of MethodView + Blueprint, sadly I don't have any examples. We'll see what davidism says.

**Comment by davidism** [comment]
Registering any type of route on a blueprint is done in the exact same way as registering on an app. `bp.route`, `bp.add_url_rule`, `bp.errorhandler`, etc.

**Comment by dougthor42** [comment]
Thanks @davidism. I don't know why I didn't know that...

Anyway, here's a working example @dfeinzeig. Note that it's just for demo purposes only - I'm sure there are some serious best practice faux pas in there.

```python
from flask import Blueprint
from flask import Flask
from flask import jsonify
from flask.views import MethodView

api = Blueprint("api", __name__)

FAKE_DATA = [{"data_id": 1, "value": "duck"}, {"data_id": 2, "value": "cat"}]


class Data(MethodView):
    def get(self, data_id):
        if data_id is None:
            # return all data
            return jsonify(FAKE_DATA)
        else:
            # Just return the one element. Ignore the selection logic here...
            return jsonify(FAKE_DATA[data_id - 1])

    def put(self):
        # fake data instead
        # data = request.get_json()
        id_ = len(FAKE_DATA) + 1
        new_data = {"data_id": id_, "value": "bird"}

        FAKE_DATA.append(new_data)
        return jsonify(new_data)


if __name__ == "__main__":
    app = Flask(__name__)

    api_view = Data.as_view("api")
    api.add_url_rule(
        "/api/v1/data/", defaults={"data_id": None}, view_func=api_view, methods=["GET"]
    )
    api.add_url_rule("/api/v1/data/<int:data_id>", view_func=api_view, methods=["GET"])
    api.add_url_rule("/api/v1/data/", view_func=api_view, methods=["PUT"])
    app.register_blueprint(api)
    app.run()
```

Testing:

```shell
$ curl localhost:5000/api/v1/data/
[{"data_id":1,"value":"duck"},{"data_id":2,"value":"cat"}]
$ curl localhost:5000/api/v1/data/1
{"data_id":1,"value":"duck"}
$ curl -X PUT localhost:5000/api/v1/data/
{"data_id":3,"value":"bird"}
$ curl localhost:5000/api/v1/data/
[{"data_id":1,"value":"duck"},{"data_id":2,"value":"cat"},{"data_id":3,"value":"bird"}]
```

**Comment by dfeinzeig** [comment]
thank you!
